### PR TITLE
Enemy/attack

### LINKED ISSUE #8: Enemy Attack
Implement enemy shooting to our player, or using melee attacks.

### PR DESCRIPTION
**Pull request by BernardoCMoreira**
RAYCAST : Enemies can attack player, and player can attack enemies. Feedback from shooting is not provided (No flash nor particles after shooting). 

- [x] Player shoot
- [x] Enemy shoot
- [x] Basic health system 
- [x] Enemy melee attack setup (damage, etc)
- [x] Basic weapon system
- [x] Enemy switching attack modes (melee/shoot)

Closes #8

### CODE REVIEW

File: Assets/Scripts/Characters/Shooter.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+using UnityEngine;
+
+/* Player and enemies should extend this */
+public abstract class Shooter : MonoBehaviour
+{
+    [SerializeField] protected float fireRate;
+    protected float timeToFire;
+    [SerializeField] protected Camera cam;
```
Comment (CajoAlbuquerque) [inline Assets/Scripts/Characters/Shooter.cs]: Why is the camera needed here? It seems only the PlayerShooterController is using, so I think it is better to move the cam variable to that class

File: Assets/Scripts/Characters/Shooter.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+using UnityEngine;
+
+/* Player and enemies should extend this */
+public abstract class Shooter : MonoBehaviour
+{
+    [SerializeField] protected float fireRate;
+    protected float timeToFire;
```
Comment (CajoAlbuquerque) [inline Assets/Scripts/Characters/Shooter.cs]: While this isn't wrong, would you say that different weapons would have different fire rates? If so, the fire rate and the whole shooting condition should be moved to the Weapon class.  
What do you think?

File: Assets/Scripts/Enemy/Controller/EnemyAI.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                if (hit.transform.name == "Player")
+                {
+                    Debug.Log("Enemy Shot");
+
+                    Player _playerStat = _player.GetComponent<Player>();
+                    _playerStat.getHealth().TakeDamage(weapon.getDamage());
+                }
+            }
+        }
```
Comment (CajoAlbuquerque) [inline Assets/Scripts/Enemy/Controller/EnemyAI.cs]: This loop is only executing once. Maybe its the `fireRate` and `timeToFire` variables which start as 0?  
I guess in line 60 we should decrease the `timeToFire` value, which is not happening...

File: Assets/Scripts/Enemy/Controller/EnemyAI.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+            if (distance <= _agent.stoppingDistance)
+            {
+                faceTarget(); /*face target*/
+                meleeAttack();
+            }
+            else
+            {
+                shoot();
+            }
```
Comment (CajoAlbuquerque) [inline Assets/Scripts/Enemy/Controller/EnemyAI.cs]: Shouldn't the `faceTarget()` be called on both cases? Doesn't the enemy need to face the player in order for the raycast to hit?

File: Assets/Scripts/Enemy/Controller/EnemyAI.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+{
+    private NavMeshAgent _agent;
+    private Transform _target;
+    private GameObject _player;
+    [SerializeField] float lookRadius = 10f;
+        void Start()
+    {
+        _player = GameObject.Find("Player"); /* Make this a singleton + game manager*/
+        _target = _player.transform; /* Make this a singleton + game manager*/
+        _agent = GetComponent<NavMeshAgent>();
```
Comment (CajoAlbuquerque) [inline Assets/Scripts/Enemy/Controller/EnemyAI.cs]: Why would this need to be a singleton? Just curious really, not opposing to it ;)  
Btw, could you pls format the document? Line 12 is not formatted

File: Assets/Scripts/Player/Controller/PlayerShootController.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            timeToFire = Time.time + 1f / fireRate;
+            RaycastHit hit;
+            if (Physics.Raycast(cam.transform.position, cam.transform.forward, out hit, weapon.getRange()))
+            {
+                if (hit.transform.name == "Enemy")
+                {
+                    Debug.Log("Player Shot");
+                    Enemy enemy = hit.transform.gameObject.GetComponent<Enemy>();
+                    enemy.getHealth().TakeDamage(weapon.getDamage());
```
Comment (CajoAlbuquerque) [inline Assets/Scripts/Player/Controller/PlayerShootController.cs]: We should add a `TakeDamage()` method to both the `Enemy` and  the `Player` classes to make it cleaner. The method would only need to call `_healt.TakeDamage()`

File: Assets/Scripts/Characters/Shooter.cs (hunk 2 above)
Comment (claudiarmamede) [reply Assets/Scripts/Characters/Shooter.cs]: Yup, I agree...  Moved both to Weapon but not sure if timeToFire will not be the same for every weapon... either way, I think they're fine in Weapon

File: Assets/Scripts/Enemy/Controller/EnemyAI.cs (hunk 4 above)
Comment (BernardoCMoreira) [reply Assets/Scripts/Enemy/Controller/EnemyAI.cs]: Yes you are right! Thanks for the warning

File: Assets/Scripts/Enemy/Controller/EnemyAI.cs (hunk 3 above)
Comment (BernardoCMoreira) [reply Assets/Scripts/Enemy/Controller/EnemyAI.cs]: Not really, i think this is called more than once. 
We are setting a new time to fire, and when that time is reached, the enemy shoots and a new time to fire is set

**Review by CajoAlbuquerque — DISMISSED** [review]
I have reviewed the latest changes and everything is much better! :D  
Are you still gonna implement melee attacks on this PR?

**Review by CajoAlbuquerque — APPROVED** [review]

**Pull request merged**